### PR TITLE
Support dependencies on multiple variants of the same component

### PR DESCRIPTION
**Pull request by melix**
### Context

This pull request adds support for having multiple dependencies to the same component, but on different variants. Previously to this change, if 2 dependencies were found on the same component but that the attributes didn't match, then we would fail. This means that this PR doesn't introduce a breaking change, but rather improves on an error case. It's worth noting that after initial review feedback, this behavior is **only allowed if variants provide different capabilities**.

This is the first step towards supporting "optional dependencies" as variants.

### CODE REVIEW

**Comment by ljacomet** [comment]
@bigdaz Would you mind having a look as well?
I did a pass and LGTM but I would feel better with a second pair of eyes confirming.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java [hunk 1]
```diff
@@ -375,6 +386,11 @@ public boolean isSelected() {
         return !incomingEdges.isEmpty();
     }
 
+    public void exclude() {
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java]: done

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/MultipleVariantSelectionIntegrationTest.groovy [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+import org.gradle.integtests.fixtures.GradleMetadataResolveRunner
+import org.gradle.integtests.fixtures.RequiredFeature
+import org.gradle.integtests.fixtures.RequiredFeatures
+import org.gradle.integtests.resolve.AbstractModuleDependencyResolveTest
+import spock.lang.Unroll
+
+@RequiredFeatures(
+        @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
+)
+class MultipleVariantSelectionIntegrationTest extends AbstractModuleDependencyResolveTest {
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/MultipleVariantSelectionIntegrationTest.groovy]: Test coverage added

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/DependenciesAttributesIntegrationTest.groovy [hunk 3]
```diff
@@ -529,54 +529,7 @@ class DependenciesAttributesIntegrationTest extends AbstractModuleDependencyReso
     }
 
     @RequiredFeatures(
-        @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "true")
-    )
-    def "Fails resolution because dependency attributes and constraint attributes conflict"() {
```
[the remark this answers is not in the document]
Comment (melix) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/attributes/DependenciesAttributesIntegrationTest.groovy]: I restored the test, it now fails with a different error (capability conflict).

**Pull request merged**